File: scripts/block/turret/prism.js
//引用部分，类似import，对应的是exports导出
const lib = require("blib");
const items = require("game/items");
//三棱镜光线聚集的强度
const angleShiftStrength = 7;
//角度
const shiftAngel = 55;
const sideOffset = 3;
//激光粗度变化速度，<1
const fade = 0.012;
//射击间隔
const shootDuration = 480;
//烎移动炮台速度
const firingMoveFract = 0.25;
//reload还要注释?
const reloadTime = 240;
//炮塔颜色变化速度
const shiftSpeed = 1.8;
var rainbowRegions = [];

const prism = extendContent(PowerTurret, "prism", {
    load(){
        this.super$load();
        for(var i = 0; i < 8; i++){
            rainbowRegions[i] = Core.atlas.find("btm-prism-rainbow-" + (i + 1));
        }
    },
    setStats() {
        this.super$setStats();
        this.stats.remove(Stat.booster);
        this.stats.add(Stat.input, StatValues.boosters(reloadTime, this.consumes.get(ConsumeType.liquid).amount, this.coolantMultiplier, false, l => this.consumes.liquidfilters.get(l.id)));
        this.stats.remove(Stat.ammo);
        this.stats.add(Stat.ammo, Core.bundle.format("bullet.damage", "540 (× 6shoots) / s"));
    },
});
prism.buildType = prov(() => {
    var bullet = [null, null, null, null, null, null];
    var bulletLife = 0;
    var bulletTime = 0;
    var _reload = reloadTime;
    const tr = new Vec2();
    const tr2 = new Vec2();
    const block = prism;
    var x=0,y=0;
    var rotation =0;
    var recoil = 0;
    var liquids = null;
    return new JavaAdapter(PowerTurret.PowerTurretBuild, {
        _bulletHeat: 0,
        getBulletHeat(){
            return this._bulletHeat;
        },
        draw(){
            rotation = this.rotation;
            recoil = this.recoil;
            this.super$draw();
            tr2.trns(rotation, -recoil);
            Draw.blend(Blending.additive);
            for(var h = 0; h < 8; h++){
                Draw.color(Color.valueOf("ff0000").shiftHue((Time.time * shiftSpeed) + (h * (360 / 8))));
                Draw.rect(rainbowRegions[h], x + tr2.x, y + tr2.y, rotation - 90);
            }
            Draw.blend();
            Draw.color();
        },
        updateTile() {
            x = this.x;
            y = this.y;
            rotation = this.rotation;
            liquids = this.liquids;
            this.super$updateTile();
            if (bulletLife > 0 && bullet[1] != null) {
                this.wasShooting = true;
                for(var n = 0; n < bullet.length; n++){
                    var data = (n * (360 / bullet.length));
                    var sine = Mathf.sinDeg(data + (bulletTime * angleShiftStrength));
                    tr.trns(rotation, (block.size * Vars.tilesize / 2) + -block.recoilAmount, sine * sideOffset);
                    bullet[n].rotation(rotation + ((sine * shiftAngel) * this._bulletHeat));
                    bullet[n].set(x + tr.x, y + tr.y);
                    bullet[n].time = 0;
                }
                this.heat = 1;
                this.recoil = block.recoilAmount;
                bulletTime += Time.delta;
                bulletLife -= Time.delta / Math.max(this.efficiency(), 0.00001);
                if (bulletLife <= 0) {
                    for(var b = 0; b < bullet.length; b++){
                        bullet[b] = null;
                    }
                    bulletTime = 0;
                    this._bulletHeat = 0;
                }
            } else if (_reload > 0) {
                this.wasShooting = true;
                var liquid = liquids.current();
                var maxUsed = block.consumes.get(ConsumeType.liquid).amount;

                var used = (this.cheating() ? maxUsed * Time.delta : Math.min(liquids.get(liquid), maxUsed * Time.delta)) * (1 + (liquid.heatCapacity - Liquids.water.heatCapacity) * block.coolantMultiplier);
                _reload -= used;
                liquids.remove(liquid, used);

                if (Mathf.chance(0.06 * used)) {
                    block.coolEffect.at(x + Mathf.range(block.size * Vars.tilesize / 2), y + Mathf.range(block.size * Vars.tilesize / 2));
                }
            }
            this._bulletHeat = Mathf.lerpDelta(this._bulletHeat, 0, fade);
        },
        updateShooting(){
        
            if(bulletLife > 0 && bullet[1] != null){
                return;
            }
            if(_reload <= 0 && (this.consValid() || this.cheating())){
                var type = this.peekAmmo();
                this.shoot(type);
            
                _reload = reloadTime;
            }
        },
    
        bullet(type, angle){
            bulletTime = 0;
            bulletLife = shootDuration;
            for(var s = 0; s < 6; s++){
                var data = (s * (360 / 6));
                var sine = Mathf.sinDeg(data + (bulletTime * angleShiftStrength));
                tr.trns(angle, block.size * Vars.tilesize / 2, sine * sideOffset);
                bullet[s] = type.create(this.tile.build, this.team, x + tr.x, y + tr.y, angle + (sine * shiftAngel));
                bullet[s].data = data;
            }
        
            this._bulletHeat = 1;
        },
    
        turnToTarget(targetRot) {
            this.rotation = Angles.moveToward(this.rotation, targetRot, this.efficiency() * block.rotateSpeed * this.delta() * (bulletLife > 0 ? firingMoveFract : 1));
        },
    
        shouldActiveSound(){
            return bulletLife > 0 && bullet[1] != null;
        },
        write(write){
            this.super$write(write);
            write.f(_reload);
        },
        read(read, revision){
            this.super$read(read, revision);
            _reload = read.f();
        },
    }, prism);
});
prism.shootEffect = Fx.shootBigSmoke2;
prism.shootCone = 40;
prism.recoilAmount = 4;
prism.size = 5;
prism.shootShake = 2;
prism.range = 29 * 8;
prism.reloadTime = reloadTime;
prism.powerUse = 34;
prism.shootSound = lib.loadSound("prismS");
prism.loopSound = lib.loadSound("prism-beam");
prism.loopSoundVolume = 2;
prism.canOverdrive = false;

prism.shootType = (() => {
    const width = 8;
    const strokes = [1.2, 1.1, 1, 0.9];
    const tscales = [1, 0.74, 0.5, 0.24];
    const lenscales = [0.92, 1, 1.017, 1.025];
    const length = 32 * 8;
    const spaceMag = 35;
    const oscMag = 1.5;
    const oscScl = 0.8;
    const cl = new JavaAdapter(ContinuousLaserBulletType, {
        draw(b){
            var fout = Mathf.clamp(b.time > b.lifetime - this.fadeTime ? 1 - (b.time - (this.lifetime - this.fadeTime)) / this.fadeTime : 1);
            var wide = Mathf.clamp(1 - b.owner.getBulletHeat());
        
            Draw.blend(Blending.additive);
            for(var s = 0; s < 4; s++){
                Draw.color(Color.valueOf("ff0000").shiftHue((Time.time * 2) + b.data));
                for(var i = 0; i < tscales.length; i++){
                    Tmp.v1.trns(b.rotation() + 180, (lenscales[i] - 1) * spaceMag);
                    Lines.stroke((width + Mathf.absin(Time.time, oscScl, oscMag)) * fout * wide * strokes[s] * tscales[i]);
                    Lines.lineAngle(b.x + Tmp.v1.x, b.y + Tmp.v1.y, b.rotation(), length * lenscales[i], false);
                }
            }
            Draw.reset();
            Draw.blend();
         },
    });
    cl.damage = 40;
    cl.hitEffect = new Effect(16, cons(e => {
        Draw.blend(Blending.additive);
        Draw.color(Color.valueOf("ff0000ff").shiftHue(Time.time * 2.0));
        Lines.stroke(e.fout() * 1.5);
        const hl = new Floatc2({get: function(x, y){
            const ang = Mathf.angle(x, y);
            Lines.lineAngle(e.x + x, e.y + y, ang, e.fout() * 8 + 1.5);
            }});
        Angles.randLenVectors(e.id, 1, e.finpow() * 70.0, e.rotation, 80.0, hl);
        Draw.blend();
        Draw.reset();
    }));
    cl.drawSize = 380;
    cl.incendChance = 0;
    return cl;
})();
prism.health = 200 * 5 * 5;
prism.coolantMultiplier = 1;
prism.consumes.add(new ConsumeCoolant(0.7)).update = false;
prism.requirements = ItemStack.with(
    Items.lead, 620,
    Items.silicon, 550,
    Items.metaglass, 430,
    Items.thorium, 385,
    items.lightninAlloy, 180
);
prism.buildVisibility = BuildVisibility.shown;
prism.category = Category.turret;
exports.prism = prism;

File: src/ExtraUtilities/worlds/blocks/turret/Prism.java
package ExtraUtilities.worlds.blocks.turret;

import ExtraUtilities.content.EUStatusEffects;
import ExtraUtilities.worlds.entity.bullet.PrismLaser;
import arc.math.Angles;
import arc.math.Mathf;
import arc.math.geom.Vec2;
import arc.struct.ObjectMap;
import arc.util.Time;
import arc.util.io.Reads;
import arc.util.io.Writes;
import mindustry.content.Liquids;
import mindustry.content.StatusEffects;
import mindustry.entities.bullet.BulletType;
import mindustry.gen.Bullet;
import mindustry.gen.Sounds;
import mindustry.type.StatusEffect;
import mindustry.world.blocks.defense.turrets.Turret;
import mindustry.world.consumers.Consume;
import mindustry.world.consumers.ConsumeLiquid;
import mindustry.world.consumers.ConsumeLiquidBase;
import mindustry.world.meta.Stat;
import mindustry.world.meta.StatValues;

import java.util.Arrays;

import static mindustry.Vars.*;

/**
 * @author guiY
 * redo 2024 0307
 * 重新做了什么？持续发射持续消耗；子弹显示子弹效果子弹伤害
 * */

public class Prism extends Turret {
    //三棱镜光线聚集的强度
    public float angleShiftStrength = 5;
    //角度
    public float shiftAngel = 30;
    //初始散射和最后的交替散射
    public float sideOffsetFirst = 4;
    public float sideOffset = 3f;
    //激光粗度和交替变化速度，<1
    public float fade = 0.001f;
    //纯交替速度（越大越慢
    public int upTime = 3;
    //烎移动炮台速度
    public float firingMoveFract = 0.4f;
    //还原泰拉瑞亚棱镜的射击最高持续时间
    public float maxShootDr = 60 * 60f;

    public StatusEffect[] statuses = {
            StatusEffects.blasted,
            StatusEffects.shocked,
            EUStatusEffects.poison,
            StatusEffects.freezing,
            StatusEffects.wet,
            StatusEffects.sapped
    };
    public BulletType shootType;
    public BulletType[] shootTypes = {};
    public float rainbowDamage = 90;
    public float rainbowWidth = 11;

    public Prism(String name) {
        super(name);
        canOverdrive = false;
        coolantMultiplier = 1f;
        shootSound = Sounds.laser;//EUSounds.prismShoot;
        loopSound = Sounds.pulse;//EUSounds.prismLoop;
        loopSoundVolume = 2f;
        recoilTime = 30f;

        consumeLiquid(Liquids.cryofluid, 30/60f);
        liquidCapacity = 30f;
    }

    @Override
    public void init() {
        super.init();
        if(statuses.length > 0) {
            shootTypes = new BulletType[statuses.length];

            for (int i = 0; i < statuses.length; i++) {
                int finalI = i;
                shootTypes[i] = new PrismLaser(rainbowDamage, rainbowWidth, range){{
                    status = statuses[finalI];
                    statusDuration = 60f;

                    if(finalI < 3) pierceArmor = true;
                }};
            }
            shootType = shootTypes[0];
        }
    }

    @Override
    public void setStats() {
        super.setStats();
        this.stats.remove(Stat.reload);
        this.stats.remove(Stat.inaccuracy);

        for(int i = 0; i < shootTypes.length; i++) {
            this.stats.add(Stat.ammo, StatValues.ammo(ObjectMap.of(new Object[]{statuses[i], shootTypes[i]})));
        }
    }

    public class PrismBuild extends TurretBuild{
        public Bullet[] bullets = new Bullet[statuses.length];

        public float bulletTime = 0f;
        public float bulletHeat = 0f;
        public float faded = fade;
        public float bulletLife = 0f;

        public Vec2 tr = new Vec2();

        public BulletType useAmmo() {
            return Prism.this.shootType;
        }

        public boolean hasAmmo() {
            return canConsume();
        }

        private boolean liquidAcc(){
            return !hasLiquids || liquids.get(liquids.current()) > liquidCapacity/6;
        }

        @Override
        public boolean canConsume() {
            return super.canConsume() && liquidAcc();
        }

        public BulletType peekAmmo() {
            return Prism.this.shootType;
        }

        @Override
        public boolean shouldConsume() {
            return isActive() || isShooting() && liquidAcc();
        }

        public boolean checkBullet(){
            if(shootTypes.length == 0 || bullets.length == 0) return false;
            for (Bullet b : bullets) {
                if (b == null) return false;
            }
            return true;
        }

        public void removeBullet(){
            for(int i = 0; i < bullets.length; i++){
                if(bullets[i] == null) continue;
                if(!bullets[i].isAdded() || bullets[i].type == null || bullets[i].owner != this) bullets[i] = null;
            }
        }

        @Override
        public void updateTile() {
            super.updateTile();
            float ammoFact = this.efficiency;
            Consume cs = Prism.this.findConsumer((f) -> f instanceof ConsumeLiquidBase);
            if (cs instanceof ConsumeLiquid cl) {
                ammoFact = Math.min(ammoFact, this.liquids.get(cl.liquid) / Prism.this.liquidCapacity);
            }

            this.unit.ammo((float)this.unit.type().ammoCapacity * ammoFact);

            removeBullet();
            if(bulletLife < maxShootDr && checkBullet()){
                wasShooting = true;
                faded = Math.min(faded + fade * Time.delta, 1f);
                for(int n = 0; n < bullets.length; n++){
                    int data = (n * (360 / bullets.length));
                    float sin = Mathf.sinDeg(data + (bulletTime * (angleShiftStrength * (1.6f - bulletHeat))));
                    tr.trns(rotation, (block.size * tilesize / 2f) - recoil + shootX, sin * (sideOffsetFirst * bulletHeat + sideOffset));
                    bullets[n].rotation(rotation + ((sin * shiftAngel) * bulletHeat));
                    bullets[n].set(x + tr.x, y + tr.y);
                    //Mana Sickness
                    bullets[n].damage = rainbowDamage * (1 - (bulletLife/maxShootDr) * 0.75f);
                    //need half -> half
                    //bullets[n].damage = rainbowDamage * Math.max(1 - bulletLife/maxShootDr, 0.5f);
                    if(isShooting() && hasAmmo()) bullets[n].time = 0;
                }
                if(isShooting() && hasAmmo()) bulletLife += Time.delta;
                curRecoil = 1f;
                heat = 1f;
                bulletTime += Time.delta;
            }
            bulletHeat = Mathf.lerpDelta(bulletHeat, 0, (float) Math.pow(faded, upTime));
        }

        @Override
        protected void updateReload() {
        }

        @Override
        protected void updateShooting() {
            if(!checkBullet()) {
                if (canConsume() && !charging() && shootWarmup >= Prism.this.minWarmup) {
                    shoot();
                }
            }
        }

        @Override
        protected void updateCooling() {
        }

        protected void shoot() {
            tr.trns(rotation, shootX, 0);
            bullet(rotation);
        }

        protected void bullet(float angle){
            Arrays.fill(bullets, null);
            bulletTime = 0;
            bulletHeat = 0;
            bulletLife  = 0;
            faded = fade;
            for(int s = 0; s < shootTypes.length; s++){
                if(shootTypes[s] == null) return;
                int data = (s * (360 / shootTypes.length));
                float sine = Mathf.sinDeg(data + (bulletTime * angleShiftStrength));
                tr.trns(angle, block.size * tilesize / 2f, sine * sideOffset);
                bullets[s] = shootTypes[s].create(tile.build, team, x + tr.x, y + tr.y, angle + (sine * shiftAngel));
                bullets[s].data = data;
            }
            bulletHeat = 1;
        }

        protected void turnToTarget(float targetRot) {
            this.rotation = Angles.moveToward(this.rotation, targetRot, power.status * Prism.this.rotateSpeed * this.delta() * (checkBullet() ? Prism.this.firingMoveFract : 1.0F));
        }

        public boolean shouldActiveSound() {
            return bullets[0] != null;
        }
    }
}
